### ISSUE TITLE
Add support for deftype

### ISSUE THREAD

**Issue by cldwalker** [babashka/sci#345]
Hi. Thanks again for defprotocol landing! Similar to #279, I figured I'd start a list of libraries that are dependent on deftype. Known libraries dependent on deftype:
* https://github.com/aysylu/loom
  * See https://github.com/borkdude/sci/issues/279#issuecomment-640283554 for more
* https://github.com/layerware/hugsql - See [these deftypes](https://github.com/layerware/hugsql/search?q=deftype&type=)
* https://github.com/clojure/data.priority-map
  * With [this bb branch](https://github.com/cldwalker/babashka/tree/data-priority-map), this library would be bb compatible if it weren't for [deftype PersistentPriorityMap](https://github.com/clojure/data.priority-map/blob/2adf0d7efb2be75b936bf2ff1e3e400d74ab64aa/src/main/clojure/clojure/data/priority_map.clj#L265)
* https://github.com/Engelberg/instaparse - Fails on [this line](https://github.com/Engelberg/instaparse/blob/4d1903b059e77dc0049dfbc75af9dce995756148/src/instaparse/auto_flatten_seq.cljc#L87)
* https://github.com/clojure/core.cache - depends on data.priority-map
* https://github.com/clojure/core.memoize - depends on core.cache and thus data.priority-map
* https://github.com/clojure/tools.reader
* https://github.com/amperity/envoy
* https://github.com/amperity/vault-clj - vault.env depends on envoy
* https://github.com/ont-app/igraph
* https://github.com/wotbrew/relic
* ~~https://github.com/clojure/test.check~~
   * ~~See https://github.com/borkdude/sci/issues/279#issuecomment-640283554 for more~~
* https://github.com/jimpil/fudje
* https://github.com/plumatic/schema/blob/2b864ab32ddf48bebe779876506577838a32dd1b/src/cljc/schema/utils.cljc#L91
* https://github.com/tonsky/datascript/search?q=deftype - for nbb

---
**Is your feature request related to a problem? Please describe.**
Would like to use libraries that depend on deftype

**Describe the solution you'd like**
Deftype compatibility for sci and thus babashka

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

**Comment by cldwalker** [comment]
Updated list above

**Comment by borkdude** [comment]
@GreshamDanielStephens Feedback on https://github.com/borkdude/sci/pull/547

- `:sci.impl/type`: I would call this :sci.impl/deftype to not conflict with other usages of this keyword
- The dot-sym-fn I would handle in the interop code. Records are also handled there, so `(Foo. x y)` works also for records
- We need to add support for field access like `(.-x (Foo. 1))` although you could access the fields through a protocol as well 
- Start adding unit tests. E.g. `(sci/eval-string "(defprotocol GetX (get-x [this])) (deftype Foo [x] GetX (get-x [this] x)) (def f (Foo. 1)) (get-x f)")` already works, could capture this in a test.
- Add support for mutable fields

**Comment by arichiardi** [comment]
Hi there, I have run into this as well while using Hugsql. The `deftype`s are here:

https://github.com/layerware/hugsql/search?q=deftype&type=

**Comment by logseq-cldwalker** [comment]
Was looking at using datascript from nbb. After fixing a couple minor things on a fork, I ran into this formidable issue again. 
@borkdude Am I understanding correctly that with [1 interface defrecord approach you floated](https://clojurians.slack.com/archives/CLX41ASCS/p1647158686052039?thread_ts=1647129108.547719&cid=CLX41ASCS), [this deftype](https://github.com/tonsky/datascript/blob/15f3d777b0d3cc818cf2163d6a1007aa2ccfc8d9/src/datascript/query_v3.cljc#L266-L288) would be doable but [this one wouldn't](https://github.com/tonsky/datascript/blob/15f3d777b0d3cc818cf2163d6a1007aa2ccfc8d9/src/datascript/lru.cljc#L6-L15)?

**Comment by borkdude** [comment]
@logseq-cldwalker Protocols aren't really a problem, it's the Java interfaces that are the problem since you can't dynamically create types in a graalvm binary. But for the above examples to work, we would have to expose `ILookup` etc as protocols to "scripts". So far this was done by a protocol -> multimethod mapping and then patching all functions that use the protocol to make use of this multimethod instead. But maybe we can revise how we do this so we are able to expose the protocols, e.g. `ILookup` without the need for a multimethod and patching.

**Comment by borkdude** [comment]
Alternatively we could make optional bindings for nbb, like we have for babashka, so you can make a home-built version of nbb.

**Comment by borkdude** [comment]
I already created an issue for the first option: https://github.com/babashka/sci/issues/639

**Comment by logseq-cldwalker** [comment]
@borkdude Gotcha. Both options sound intriguing. The first option sounds potentially more powerful and reusable for others

**Comment by borkdude** [comment]
I added very basic support for deftype: it only supports immutable fields and basically behaves the same as records in SCI.

Will follow up with more issues for:

- supporting 1 Java interfaces on deftype and defrecord, similar to the approach taken in https://github.com/babashka/sci/pull/547
- support for mutable fields and `set!`

**Comment by logseq-cldwalker** [comment]
Exciting! 🎉

**Comment by borkdude** [comment]
@logseq-cldwalker Deftype is now supported, similar to defrecord (with the same restriction that you can't add Java interfaces currently). It'd be interesting to see which of the libs are now supported by this changes in the original post.

**Comment by borkdude** [comment]
I went through the list: unfortunately many of these libraries implement their own data structures using `deftype` which isn't yet support due to the Java interface restriction. Perhaps we can "special cased" support for that use case in the future.

But I found that https://github.com/ont-app/igraph now works (at least the require doesn't crash), but due to a weird metadata usage in the library, I had to make one change in SCI / bb. So the library should now work with bb master (and if you include spec). I think we should also make spec a built-in.
